tpm2_pytss/utils.py

```diff
@@ -226,3 +228,129 @@ def unwrap(
+        publicArea=TPMT_PUBLIC(
+            type=TPM2_ALG.ECC,
+            nameAlg=TPM2_ALG.SHA256,
+            objectAttributes=attrs,
+            authPolicy=policy,
+            parameters=TPMU_PUBLIC_PARMS(
+                eccDetail=TPMS_ECC_PARMS(
+                    symmetric=symmetric,
+                    scheme=TPMT_ECC_SCHEME(scheme=TPM2_ALG.NULL),
+                    curveID=TPM2_ECC.NIST_P256,
+                    kdf=TPMT_KDF_SCHEME(scheme=TPM2_ALG.NULL),
+                )
+            ),
+            unique=TPMU_PUBLIC_ID(ecc=TPMS_ECC_POINT(x=b"\x00" * 32, y=b"\x00" * 32)),
+        )
+    )
+    rsa_template = TPM2B_PUBLIC(
+        publicArea=TPMT_PUBLIC(
+            type=TPM2_ALG.RSA,
+            nameAlg=TPM2_ALG.SHA256,
+            objectAttributes=attrs,
+            authPolicy=policy,
+            parameters=TPMU_PUBLIC_PARMS(
+                rsaDetail=TPMS_RSA_PARMS(
+                    symmetric=symmetric,
+                    scheme=TPMT_RSA_SCHEME(scheme=TPM2_ALG.NULL),
+                    keyBits=2048,
+                ),
+            ),
+            unique=TPMU_PUBLIC_ID(rsa=b"\x00" * 256),
+        )
+    )
+    if ekalg == TPM2_ALG.ECC:
+        nvbase = 0x1C0000A
+        template = ecc_template
+    elif ekalg == TPM2_ALG.RSA:
+        nvbase = 0x1C00002
+        template = rsa_template
+    else:
+        raise ValueError(f"unsupported EK algorithm {ekalg}")
+
+    nvnonce = nvbase + 1
+    nvtemplate = nvbase + 2
+
+    indices = []
+    more = True
+    while more:
+        more, data = ectx.get_capability(
+            TPM2_CAP.HANDLES,
+            TPM2_HT.NV_INDEX << 24,
+            TPM2_PT_NV.INDEX_MAX,
+            session1=session,
+        )
+        handles = list(data.data.handles)
+        indices.extend(handles)
+
+    if nvtemplate in indices:
+        thandle = ectx.tr_from_tpmpublic(nvtemplate, session1=session)
+        tpub, _ = ectx.nv_read_public(thandle, session1=session)
+        left = tpub.nvPublic.dataSize
+        mtempl = b""
+        while left > 0:
+            off = tpub.nvPublic.dataSize - left
+            size = 0x300 if left > 0x300 else left
+            data = ectx.nv_read(thandle, size, off, session2=session)
+            mtempl = mtempl + bytes(data)
+            left = left - len(data)
+        templ, _ = TPMT_PUBLIC.unmarshal(mtempl)
+        template = TPM2B_PUBLIC(publicArea=templ)
+
+    nonce = b""
+    if nvnonce in indices:
+        nhandle = ectx.tr_from_tpmpublic(nvnonce, session1=session)
+        npub, _ = ectx.nv_read_public(nhandle, session1=session)
+        left = npub.nvPublic.dataSize
+        while left > 0:
+            off = npub.nvPublic.dataSize - left
+            size = 0x300 if left > 0x300 else left
+            data = ectx.nv_read(nhandle, size, off, session2=session)
+            nonce = nonce + bytes(data)
+            left = left - len(data)
+
+    if len(nonce) > 0 and template.publicArea.type == TPM2_ALG.RSA:
```

Comment:
This padding is only done on RSA 2048 and ECC NIST P256 per the 2.2.1.6 "TPMT_PUBLIC Calculation"
  - https://trustedcomputinggroup.org/wp-content/uploads/TCG_IWG_EKCredentialProfile_v2p4_r3.pdf

Their are other keys possible, so I don't think we want to always do this AFAICT.

Reply:
The code currently only supports RSA 2048 and NIST P256 keys (unless the TPM manufacturer has put a non-standard template in one of the NV template areas)